[TASK] Change property docblock of $contentArgumentName how to set it

sbuerk opened the pull request:
This pull-request contains two commits:

- Adding tests to cover this trait and show that implementation is broken
- Provide the change to document how to set this property correctly

As this is needed for changes in the Typo3 core, it would be nice if
a new patchlevel release could be done after approving and merging this,
so a pending core change can be reviewed and processed:
https://review.typo3.org/c/Packages/TYPO3.CMS/+/73014

Fixes: #582

lolli42 commented on src/Core/ViewHelper/Traits/CompileWithContentArgumentAndRenderStatic.php:
```diff
@@ -118,7 +121,7 @@ protected function buildRenderChildrenClosure()
     public function resolveContentArgumentName()
     {
-        if (empty($this->contentArgumentName)) {
+        if (!property_exists($this, 'contentArgumentName') || empty($this->contentArgumentName)) {
```
maybe we should split !property_exists() and empty() here, to avoid writing to a non-existing property below - if the VH does not have the property, we could fall back to a second property that is exclusively used within the trait, or just do the magic each time.
Note accessing a non-existing class property will log a deprecation note with php 8.2

sbuerk replied:
Done. Please re-review if it's okay now.

mbrodala commented on src/Core/ViewHelper/Traits/CompileWithContentArgumentAndRenderStatic.php:
```diff
@@ -66,7 +75,7 @@ public function compile(
         TemplateCompiler $compiler
     ) {
-        list ($initialization, $execution) = ViewHelperCompiler::getInstance()->compileWithCallToStaticMethod(
+        [$initialization, $execution] = ViewHelperCompiler::getInstance()->compileWithCallToStaticMethod(
```
Since [Fluid still supports PHP >= 5.5](https://github.com/TYPO3/Fluid/blob/3d8499ae81e1fda39d236072f3ef8a7ef35433b1/composer.json#L6) we cannot do this.

sbuerk replied:
How came that ? This was a non intended change and not done activly oO. Will revert this.

mbrodala commented on src/Core/ViewHelper/Traits/CompileWithContentArgumentAndRenderStatic.php:
```diff
@@ -118,19 +127,30 @@ protected function buildRenderChildrenClosure()
-                    return $this->contentArgumentName;
-                }
+        // early return already defined argument name
+        if ($this->useContentArgumentName !== null) {
```
Would a local `static` variable work, too? :thinking:

sbuerk replied:
mhhhh .....

sbuerk replied:
I thing static would introduce some nasty flaws, see: https://3v4l.org/r8C8D

mbrodala replied:
That's ... interesting. It surely makes sense somewhere but yes, too much risk here.

NamelessCoder commented:
TBF, I don't really like the direction this is going - the added property that's nearly the same as the existing property, plus the involvement of a property that may or may not exist. May I suggest instead that we adjust the phpdoc and mention that overriding the `resolveContentArgumentName` method is the way to go? That fits better with composition logic and traits, at least IMHO. Plus it's testable in implementers without also invoking the trait's methods.

It also seems excessive to me to add three full VH fixtures with the only purpose of testing the behavior of this trait.

sbuerk commented:
> TBF, I don't really like the direction this is going - the added property that's nearly the same as the existing property, plus the involvement of a property that may or may not exist. May I suggest instead that we adjust the phpdoc and mention that overriding the `resolveContentArgumentName` method is the way to go? That fits better with composition logic and traits, at least IMHO. Plus it's testable in implementers without also invoking the trait's methods.

Yes, right. I have reverted my changes in my commit and added the changed docblock of the
property to describe how to correctly set this, and removed the wrong statement to simply
copy it. Generally setting it in __constructor seems to work, at least in tests, thus adding a 4th ViewHelper fixture.

> It also seems excessive to me to add three full VH fixtures with the only purpose of testing the behavior of this trait.

It is far away from excessig. At first, there was not any coverage of this trait and for this
behavior, so the first two fixture ViewHelpers are there to demonstrate the normal resolving
behaviour which already worked but was not tested explicitly.

The 3rd or better to say 3rd and 4th are there to test/demonstrates the behaviour using the constructor or overriding the resolveContentArgumentName method and that both of them works.

Personally I would add a fifth, demonstrating that setting it in 'initializeArguments()' for example wouldn't work - but left that out, as you was already against it.

Anyway the 4th fixture should stay as they are testing behiouver which *must* be covered and should not be left out. And doing the different constalltions are only possible with different ViewHelpers. Using fixture ViewHelpers decouples this from shipped viewhelpers as it tests the trait behaviour, not a special ViewHelper - which was the goal of these tests.

mbrodala commented on src/Core/ViewHelper/Traits/CompileWithContentArgumentAndRenderStatic.php:
```diff
@@ -25,12 +25,34 @@ trait CompileWithContentArgumentAndRenderStatic
+     * $this->contentArgumentName = 'explicitlyToUseArgumentName';
+     *
+     * Another possible way would be to override the method
+     * 'resolveContentArgumentName()' and returning the name.
```
```suggestion
     * 'resolveContentArgumentName()' and return the name.
```

sbuerk replied:
done

mbrodala commented on src/Core/ViewHelper/Traits/CompileWithContentArgumentAndRenderStatic.php:
```diff
@@ -25,12 +25,34 @@ trait CompileWithContentArgumentAndRenderStatic
+     *     return 'explicitlyToUseArgumentName';
+     * }
+     *
+     * Note: Setting this through 'initializeArguments()' do
```
```suggestion
     * Note: Setting this through 'initializeArguments()' will
```

sbuerk replied:
done

mbrodala commented on src/Core/ViewHelper/Traits/CompileWithContentArgumentAndRenderStatic.php:
```diff
@@ -25,12 +25,34 @@ trait CompileWithContentArgumentAndRenderStatic
+     * }
+     *
+     * Note: Setting this through 'initializeArguments()' do
+     *       not work as it may be expected, and the other
```
```suggestion
     *       not work as expected, and other
```

sbuerk replied:
done

mbrodala commented on src/Core/ViewHelper/Traits/CompileWithContentArgumentAndRenderStatic.php:
```diff
@@ -25,12 +25,34 @@ trait CompileWithContentArgumentAndRenderStatic
+     *
+     * Note: Setting this through 'initializeArguments()' do
+     *       not work as it may be expected, and the other
+     *       methods should be avoided to override for this.
```
```suggestion
     *       methods should be avoided to override this.
```

sbuerk replied:
done

mbrodala commented on src/Core/ViewHelper/Traits/CompileWithContentArgumentAndRenderStatic.php:
```diff
@@ -25,12 +25,34 @@ trait CompileWithContentArgumentAndRenderStatic
+     *       not work as it may be expected, and the other
+     *       methods should be avoided to override for this.
+     *
+     * Following test ViewHelpers are tested and demonstrates
+     * that the setting posibillities works.
+     * @see CompileWithContentArgumentAndRenderStaticExplicitSetArgumentNameForContentInConstructorViewHelper
+     * @see CompileWithContentArgumentAndRenderStaticExplicitSetArgumentNameForContentOverriddenResolveContentArgumentNameMethodViewHelper
```
This should be dropped, since it doesn't add value to the consumer and will get outdated over time.

lolli42 commented:
ping. could we get this done?

mbrodala reviewed (approved):
Perfectly fine from my POV.

NamelessCoder commented:
Thanks for taking care of this, @sbuerk - and for being receptive to the implementation input during review ;)
